network_simulator/link.py

```diff
@@ -4,44 +4,94 @@
 """
 
 from Queue import Queue
-
-# TODO(yubo): Finish implementing the classes and Events for this.
-
+from packet import *
+from common import *
+from event import *
 
 class LinkBuffer(object):
     """
     Representation of a Link's FIFO buffer.
-
-    TODO(team): Parameters related to link congestion, so we can have dynamic
-    routing table updates.
-    TODO(yubo): Functions for whether this is full etc.
     """
 
-    def __init__(self, buffer_size_bits=None, queue=Queue()):
+    # TODO(team): Parameters related to link congestion, so we can have dynamic
+    # routing table updates.
+    # TODO(yubo): packet_exemption_times deque
+
+    def __init__(self, max_buffer_size_bits=None, queue=Queue()):
         """
-        :ivar int buffer_size_bits: maximum buffer size in bits.
+        :ivar int max_buffer_size_bits: maximum buffer size in bits.
         :ivar Queue queue: a FIFO queue of LinkBufferElements
         """
-        self.buffer_size_bits = buffer_size_bits
+        self.max_buffer_size_bits = max_buffer_size_bits
         self.queue = queue
+        self.curr_buffer_size_bits = 0 # size of current buffer in bits
 
     def __repr__(self):
         return str(self.__dict__)
 
+    def remaining(self):
+        """
+        returns remaining size in buffer in bits
+
+        not strictly necessary, push() will check remaining() condition
+        :return: remaining buffer size in bits
+        """
+        return self.max_buffer_size_bits - self.curr_buffer_size_bits
+
+    def push(self, packet, dest_device_addr, global_clock_sec):
+        """
+        :param Packet packet: Packet to push
+        :param string dest_device_addr: destination device of packet
+
+        Pushes packet onto LinkBuffer going towards dest_device_addr
+        :return: True if pushed, False if not
+        """
+        if isinstance(packet, RouterPacket):
```

Comment:
I decided that RouterPackets are always pushed and don't contribute to statistics, so no size check since always pushed. Is that correct/reasonable?